### PR TITLE
Major update

### PR DESCRIPTION
**Pull request by rimutaka**
Alec, thank you very much for the work you put into making this crate. I took your `octol` branch and added a few features:
- custom attribute name prefix
- custom text node name
- no-panic on XML parsing

I also added more tests, docs and examples. I think it warrants bumping the version to 0.4.1.

We will be using this code in a production setting, so I'm likely to come back with more fixes and improvements. Let me know if you no longer interested in maintaining this repo.

Any suggestions on what I missed or how to improve the code are welcome.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,116 @@
 # quickxml_to_serde
-Convert xml to json using quickxml and serde. HEAVLY inspired from node2object
+
+Convert XML to JSON using [quick-xml](https://github.com/tafia/quick-xml) and [serde](https://github.com/serde-rs/json). Inspired by [node2object](https://github.com/vorot93/node2object).
+
+## Usage examples
+
+#### Basic
+
+```
```
Comment (AlecTroemel) [inline README.md]: some nice rust syntax highlighting
```suggestion
```rust
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,116 @@
...
+#### Custom config
+
+The following config example changes the default behavior to:
+
+1. Treat numbers starting with `0` as strings. E.g. `007` will be `"007"`
+2. Do not prefix properties created from attributes
+3. Use `text` as the property name for values of XML text nodes where the text is mixed with other nodes
+4. Exclude empty elements from the output
+
+```
```
Comment (AlecTroemel) [inline README.md]:
```suggestion
```rust
```

**Comment by AlecTroemel** [comment]
@rimutaka thank you for this PR! Sorry for letting it sit for so long, I need to be more responsive to these sort of things. I discussed it with my employer (we're also using this in production!) and they agreed to give you write permissions to this, you should see the invite!  

Looking at our use of this crate,  we're actually using the `map_over_children` and `map_of_children` functions. I'd ask we keep those in for now, at least until we find a better place for those. also, I added some suggestions in the readme to get syntax highlighting on the code examples.

If this project continues to grow, I think it may be a good idea to split up `lib.rs` into multiple files. Pulling out the config  source would be a good place to start.. but that can wait for another day :smile:

File: src/lib.rs [hunk 3]
```diff
@@ -74,108 +200,41 @@ fn convert_node(el: &Element) -> Option<Value> {
...
     );
     Value::Object(data)
 }
 
-pub fn xml_string_to_json(xml: String) -> Value {
-    let root = Element::from_str(xml.as_str()).unwrap();
-    xml_to_map(&root)
-}
-
-pub fn map_over_children<T: BufRead, F: FnMut(&str, &Value)>(xml: T, mut iteratee: F) {
```
Comment (AlecTroemel) [inline src/lib.rs:95]: could we keep these functions in for now?

File: src/lib.rs [hunk 4]
```diff
@@ -74,108 +200,41 @@ fn convert_node(el: &Element) -> Option<Value> {
...
-        let mut child_xml = Vec::new();
-        child
-            .write_to(&mut child_xml)
-            .expect("successfully write to the vector");
-        let xml_string = String::from_utf8(child_xml).unwrap();
-        iteratee(xml_string.as_str(), &xml_to_map(&child));
-    }
-}
-
-pub fn map_of_children(root: Element) -> Vec<(String, Value)> {
```
Comment (AlecTroemel) [inline src/lib.rs:109]: this one as well

**Comment by rimutaka** [comment]
Hi @AlecTroemel . Thanks for getting back to me so quick.
1. Invitation accepted, but I will have to rely on your review and approval before any merges.
2. Will address the review comments - all good and sensible.
3. Can you share some use cases for those pub fn's I deleted? There were commented out tests. I would need to make those work as well.

**Comment by AlecTroemel** [comment]
for point 3, that's a fair question. Looking more into how I'm using those functions (`xml_to_map` and `map_over_children`), it looks very specific to my use case. Im using them to iterate over the top level children of a node. something like this

```xml
<!-- xml.file-->
<top>
   <child>...</child>
   <child>...</child>
   <child>...</child>
</top>
```

```rust
let root = match get_root("file.xml").unwrap();
let data = map_of_children(root);
for (_, child_json) in data.iter_mut() {
    // child_json will be the json of <child>...</child>
    do_something_with_child(child_json);
}
```

Because of that I'm ok deleting those 2 functions here and moving them into my project.  You've convinced me :laughing:

**Comment by rimutaka** [comment]
@AlecTroemel Alec, I made changes as requested, except for those 2 functions. Do you have a minute for another review before I merge?

**Comment by AlecTroemel** [comment]
LGTM

I think I'm good closing #1 and #2 as well with this

**Comment by AlecTroemel** [comment]
@rimutaka merged and published to crates.io! Do you have an account there? if so I can add you as an Owner so that you can publish versions as well

**Comment by rimutaka** [comment]
Choice! My crates a/c  
- Name: mx
- GitHub Account: rimutaka
- email: max@onebro.me
Not sure which one of them you need for an invite.

**Pull request merged**